cpp/src/datetime/datetime_ops.cu

```diff
@@ -223,5 +244,12 @@ std::unique_ptr<column> last_day_of_month(column_view const& column,
       (column, 0, mr);
 }
 
+std::unique_ptr<column> day_of_year(column_view const& column,
+                                    rmm::mr::device_memory_resource* mr) {
+  CUDF_FUNC_RANGE();
+  return detail::apply_datetime_op<detail::extract_day_num_of_year, cudf::INT16>
```

Comment:
I can't comment on lines you didn't change, but I think apply_datetime_op should do these two things:

1. CUDF_EXPECTS that the column has datetime type before trying to launch the op.
2. If the column is empty, return an empty column (before launching the op).